### PR TITLE
ARROW-7587: [C++][Compute] Implement nth_to_indices kernel

### PR DESCRIPTION
**Pull request by cyb70289**
Perform indices partition of an array around n-th smallest element.
Output indices[0 ~ n) holds values no greater than n-th element, and
indices[n+1 ~ end) holds values no less than n-th element. Elements
in each partition are not sorted. Nulls will be partitioned to the
end of the output. Output is not stable.

This is useful for picking top-n elements from an array. It leverages
O(n) std::nth_element which is orders faster than std::sort. See Jira
link for benchmark results of nth_element, sort, and partial_sort:
https://issues.apache.org/jira/browse/ARROW-7587

### CODE REVIEW

**Comment by cyb70289** [comment]
Some code duplicated with sort_to_indices.cc, as show in appveyor failure: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/31228327/job/72i22arf2cu8q8s4#L1031
Shall I extract common code to a new file?

**Comment by pitrou** [comment]
You can indeed create a file  `sort_internal.h` or something.

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+class ARROW_EXPORT NthToIndicesKernel : public OpKernel {
+ protected:
+  std::shared_ptr<DataType> type_;
+
+ public:
+  /// \brief Kernel interface
+  ///
+  /// delegates to subclasses via NthToIndices()
+  virtual Status Call(FunctionContext* ctx, const Datum& values, uint64_t n,
+                      Datum* offsets) = 0;
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices.cc]: Is there a point in having virtual methods? Are you planning to create a second implementation of this kernel?

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    int64_t buf_size = values->length() * sizeof(uint64_t);
+    RETURN_NOT_OK(AllocateBuffer(ctx->memory_pool(), buf_size, &indices_buf));
+
+    int64_t* indices_begin = reinterpret_cast<int64_t*>(indices_buf->mutable_data());
+    int64_t* indices_end = indices_begin + values->length();
+
+    std::iota(indices_begin, indices_end, 0);
+    *offsets = std::make_shared<UInt64Array>(values->length(), indices_buf);
+
+    if (n >= static_cast<uint64_t>(values->length())) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices.cc]: Shouldn't we have two cases here?
* `n` is exactly equal to the length: ok
* `n` strictly greater than the length: return `Status::Invalid`

File: cpp/src/arrow/compute/kernels/nth_to_indices.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+/// holds values no less than n-th element. Elements in each partition
+/// is not sorted. Nulls will be partitioned to the end of the output.
+/// Output is not guaranteed to be stable.
+///
+/// \param[in] ctx the FunctionContext
+/// \param[in] values array to sort
+/// \param[in] n n-th element to partition the array
+/// \param[out] offsets indices that would sort an array
+ARROW_EXPORT
+Status NthToIndices(FunctionContext* ctx, const Array& values, uint64_t n,
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices.h]: I know `n` shouldn't be negative, but our convention is to use `int64_t` for lengths and offsets, so we should probably use that here too.

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  /// \param[in] value_type constructed NthToIndicesKernel will support
+  ///            values of this type
+  /// \param[out] out created kernel
+  static Status Make(const std::shared_ptr<DataType>& value_type,
+                     std::unique_ptr<NthToIndicesKernel>* out);
+};
+
+template <typename ArrayType>
+bool CompareValues(const ArrayType& array, uint64_t lhs, uint64_t rhs) {
+  return array.Value(lhs) < array.Value(rhs);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices.cc]: Note that `GetView` also exists on numeric arrays, so you don't need two different functions.

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      kernel = MakeNthToIndicesKernelImpl<StringType>(CompareViews<StringArray>);
+      break;
+    default:
+      return Status::NotImplemented("N-th of ", *value_type, " arrays");
+  }
+  out->reset(kernel);
+  return Status::OK();
+}
+
+Status NthToIndices(FunctionContext* ctx, const Datum& values, uint64_t n,
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices.cc]: Do you want to expose this function in the `.h`?

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+#include "arrow/type_traits.h"
+
+namespace arrow {
+
+class Array;
+
+namespace compute {
+
+/// \brief Kernel implementing NthToIndices operation
+class ARROW_EXPORT NthToIndicesKernel : public OpKernel {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices.cc]: By the way, all private code should be enclosed in the anonymous namespace (and you can remove `ARROW_EXPORT`).

File: cpp/src/arrow/compute/kernels/nth_to_indices_test.cc [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      }
+    }
+  }
+
+ protected:
+  void AssertNthToIndicesArray(const std::shared_ptr<Array> values, int n) {
+    std::shared_ptr<Array> offsets;
+    ASSERT_OK(arrow::compute::NthToIndices(&this->ctx_, *values, n, &offsets));
+    ASSERT_OK(offsets->ValidateFull());
+    ValidateNth<ArrayType>(*std::static_pointer_cast<ArrayType>(values), n,
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/nth_to_indices_test.cc]: Can use `checked_pointer_cast` from `util/checked_cast.h`.

File: cpp/src/arrow/compute/kernels/nth_to_indices.h (hunk 3 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/nth_to_indices.h]: Something else: you can return `Result<std::shared_ptr<Array>>` instead of Status + out-pointer.

**Review by pitrou — CHANGES_REQUESTED** [review]
Looks good generally, just a couple comments. Thank you for doing this!

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices.cc]: removed virtual

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc (hunk 2 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices.cc]: yes. done.

File: cpp/src/arrow/compute/kernels/nth_to_indices.h (hunk 3 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices.h]: Changed n to int64.
I'm returning Status just to be consistent with other kernels. Result looks more convenient.

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc (hunk 4 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices.cc]: Thanks. It simplifies code significantly. I think sorting kernel should also be refined with it.

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices.cc]: no, changed to static.

File: cpp/src/arrow/compute/kernels/nth_to_indices.cc (hunk 6 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices.cc]: removed ARROW_EXPORT.
I'm kind of confused about how compute kernels are used by client code.
Exported functions like SortToIndices will create a kernel every time client invokes it. Looks to me exporting the kernel class interface is better, so clients can create a kernel, then call it against different inputs repeatedly.

File: cpp/src/arrow/compute/kernels/nth_to_indices_test.cc (hunk 7 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/nth_to_indices_test.cc]: done

**Comment by cyb70289** [comment]
Hi @pitrou , I have updated this patch. Please help review when you have time. Thanks.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @cyb70289 . I pushed a minor change and will merge when green.

**Pull request closed without merging**